### PR TITLE
feat(list): Add arrow key a11y support. 

### PR DESCRIPTION
**Pull request by williamernest**
fixes: #2850

### CODE REVIEW

File: packages/mdc-list/foundation.js [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+      } else {
+        // Return early because first item is already focused.
+        return;
+      }
+    }
+    this.adapter_.focusItemAtIndex(prevIndex);
+  }
+
+  focusFirstElement() {
+    this.adapter_.focusItemAtIndex(0);
```
Comment (moog16) [inline packages/mdc-list/foundation.js]: do you need a check here to make sure there `this.adapter_.getListItemCount() > 0` ?

File: packages/mdc-list/foundation.js [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  }
+
+  /**
+   * Utility method to find the first ancestor with the mdc-list-item class.
+   * @param {EventTarget} target
+   * @return {?Element}
+   * @private
+   */
+  getListItem_(target) {
+    while (!target.classList.contains('mdc-list-item')) {
```
Comment (moog16) [inline packages/mdc-list/foundation.js]: I think 'mdc-list-item' should be in strings

File: packages/mdc-list/index.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import MDCComponent from '@material/base/component';
+import {MDCListFoundation} from './foundation';
+import {strings} from './constants';
+
+/**
+ * @extends MDCComponent<!MDCistFoundation>
```
Comment (moog16) [inline packages/mdc-list/index.js]: MDCistFoundation => MDCListFoundation

File: packages/mdc-list/index.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    this.root_.addEventListener('focusin', this.focusInEventListener_);
+    this.root_.addEventListener('focusout', this.focusOutEventListener_);
+    this.layout();
+  }
+
+  layout() {
+    const direction = this.root_.getAttribute(strings.ARIA_ORIENTATION);
+    this.vertical = direction === strings.ARIA_ORIENTATION_VERTICAL;
+
+    // List items need to have at least tabindex=-1 to be focusable.
```
Comment (moog16) [inline packages/mdc-list/index.js]: you mean tabindex=0?

File: packages/mdc-list/README.md [hunk 5]
```diff
@@ -137,45 +135,96 @@ OR
...
+when the list item is not focused. When the list item receives focus, the child `button` and `a` elements will 
+receive `tabIndex="0"`. This allows for the user to tab through a list items elements and then tab to the
+first element after the list. The `Arrow`, `Home`, and `End` keys should be used for navigating internal list elements.
+The MDCList will perform the following actions for each key press
+
+Key | Action
+--- | ---
+`ArrowUp` | When the list is in a vertical orientation, it will cause the previous list item to receive focus.
+`ArrowDown` | When the list is in a vertical orientation, it will cause the next list item to receive focus.
+`ArrowLeft` | When the list is in a horizontal orientation, it will cause the previous list item to receive focus.
```
Comment (moog16) [inline packages/mdc-list/README.md]: When the list is in a horizontal orientation (default), ...

File: packages/mdc-list/README.md [hunk 6]
```diff
@@ -137,45 +135,96 @@ OR
...
+`setTabIndexForListItemChildren(ndx: Number, value: Number) => void` | Sets the `tabindex` attribute to `value` for each child `button` and `a` element in the list item at the `ndx` specified.
+
+### `MDCListFoundation`
+
+Method Signature | Description
+--- | ---
+`setWrapFocus(value: Boolean) => void` | Sets the list to allow the up arrow on the first element to focus the last element of the list and vice versa. 
+`setVerticalOrientation(value: Boolean) => void` | Sets the list to an orientation causing the keys used for navigation to change. `true` results in the Up/Down arrow keys being used. `false` results in the Left/Right arrow keys being used. 
+`handleFocusIn(evt: Event) => void` | Handles the changing of `tabindex` to `0` for all `button` and `a` elements when a list item receives focus. 
+`handleFocusOut(evt: Event) => void` | Handles the changing of `tabindex` to `-1` for all `button` and `a` elements when a list item receives focus.
```
Comment (moog16) [inline packages/mdc-list/README.md]: when a list item loses focus (blur)?

File: packages/mdc-list/index.js (hunk 4 above)
Comment (williamernest) [reply packages/mdc-list/index.js]: No, tabindex=-1. `li` elements cannot be focused unless they have a `tabIndex` value. If it's 0, then they're in the tab order. If it's -1, they're not in the tab order, but they are focusable.

**Comment by patbou02** [comment]
I am investigating a bug where pressing the up/down keys while a list is expanded will make the selected list item skip an element of the list. Is anyone familiar with this? Thank you in advance.

**Comment by RebeccaStankus** [comment]
@patbou02  We came across this same issue with the arrow keys skipping every other element in our list. We found that the MDC drawer and MDC list are both trying to control the arrow key functions, so we removed the `MDCDrawer` instantiation from our JavaScript code and the arrow keys work with our list as expected now. Hope this helps!

**Pull request merged**